c2rust-analyze/src/borrowck/type_check.rs

```diff
@@ -271,17 +272,20 @@ impl<'tcx> TypeChecker<'tcx, '_> {
                 })
             }
 
-            Rvalue::Cast(_, _, ty) => self.ltcx.label(ty, &mut |_ty| {
-                // TODO: handle Unsize casts at minimum
-                /*
-                assert!(
-                    !matches!(ty.kind(), TyKind::RawPtr(..) | TyKind::Ref(..)),
-                    "pointer Cast NYI"
-                );
-                */
-                Label::default()
-            }),
+            Rvalue::Cast(ck, _, ty) => {
+                // TODO: handle Unsize casts
 
+                // special case in casting to void*
+                if expect_ty.label.perm.contains(PermissionSet::FREE) {
+                    return expect_ty;
```

Comment:
@spernsteiner these two `if` blocks feel incorrect, but I'm not sure what the proper label would be for types casted from `void*` to `my_struct*` that are eventually passed to `free`. There are downstream checks that the origin params are equivalent in size between LHS and RHS types, and so it seems like we either have to 
- create `origin_params` out of thin air
- copy the structure from `expect_ty` or
- avoid asserting `LTy` `origin_params` length parity when casting from `void*` to `my_struct*`.

What do you think?

Reply:
I think the general approach (for both borrowck and dataflow) should be to recognize this pattern:
```
_1 = malloc(...);
_2 = _1 as *mut T;
```
and instead treat it as if the MIR code read
```
_2 = malloc(...);
```
One option for implementing this would be to run a scan over the whole function up front and record some annotations when it sees the pattern: "use this `Place` instead of the real LHS" for the calls, and "ignore this statement" for the casts.  Then check for the relevant annotations in the `Call` and `Cast` cases of both borrowck and dataflow.

Another option (which might actually be simpler overall) would be to clone the entire MIR `Body` for each function and rewrite the pattern to the replacement in-place.

Reply:
@spernsteiner I decided to go with the first option